cpp_guidelines/.clang-format

```diff
@@ -0,0 +1,107 @@
+---
+Language: Cpp
+Standard: c++17
+DisableFormat:   false
+AccessModifierOffset: -4
+AlignAfterOpenBracket: Align
+AlignConsecutiveAssignments: true
+AlignConsecutiveBitFields: true
+AlignConsecutiveDeclarations: false
+AlignConsecutiveMacros: true
+AlignEscapedNewlines: Right
+AlignOperands: true
+AlignTrailingComments: true
+AllowAllArgumentsOnNextLine: false
+AllowAllConstructorInitializersOnNextLine: false
+AllowAllParametersOfDeclarationOnNextLine: false
+AllowShortBlocksOnASingleLine: Never
+AllowShortCaseLabelsOnASingleLine: false
+AllowShortEnumsOnASingleLine: false
+AllowShortFunctionsOnASingleLine: None
+AllowShortIfStatementsOnASingleLine: Never
+AllowShortLambdasOnASingleLine: All
+AllowShortLoopsOnASingleLine: false
+AlwaysBreakAfterReturnType: None
+AlwaysBreakBeforeMultilineStrings: false
+AlwaysBreakTemplateDeclarations: Yes
+BinPackArguments: false
+BinPackParameters: false
+BitFieldColonSpacing: Both
+BreakBeforeBinaryOperators: All
+BreakBeforeBraces: Allman
+BreakBeforeTernaryOperators: true
+BreakConstructorInitializers: BeforeComma
+BreakInheritanceList: BeforeComma
+BreakStringLiterals: true
+ColumnLimit:     79
+CommentPragmas:  '^ COMMENT pragma:'
+CompactNamespaces: false
+ConstructorInitializerAllOnOneLineOrOnePerLine: true
+ConstructorInitializerIndentWidth: 4
+ContinuationIndentWidth: 4
+Cpp11BracedListStyle: true
+DeriveLineEnding: false
+DerivePointerAlignment: false
+ExperimentalAutoDetectBinPacking: false
+FixNamespaceComments: true
+IncludeBlocks:   Preserve
+IncludeCategories:
+  - Regex:           '^"(llvm|llvm-c|clang|clang-c)/'
+    Priority:        2
+    SortPriority:    0
+  - Regex:           '^(<|"(gtest|gmock|isl|json)/)'
+    Priority:        3
+    SortPriority:    0
+  - Regex:           '.*'
+    Priority:        1
+    SortPriority:    0
```

Comment:
We are not LLVM so I guess we can remove this?

Reply:
Yes, that is the long term goal. I based it on the `Microsoft` style and this is what clang-format put in by default. But I was not sure what to put in as replacement so we definitely need to discuss these rules.

Reply:
The content in here depends on the project. For a low-level lib without much dependencies, we do not need the `llvm` block. For PIConGPU on the other hand, this block should probably contain the close dependencies like `alpaka`.

Reply:
**VOTE**

| current | remove entire section |
| -- | -- |
| 0 | 1 |

I just tried removing the entire section and then clang-format just sorts the includes within their groups lexicographically. This should be fine. If we ever want to introduce groups, we can do that at a later stage. I do not think we need it now. And we definitely do not need groups for llvm, clang, gtest etc.

Reply:
current | remove entire section
-- | --
0 | 2

I agree with @bernhardmgruber the current selection of groups is not representing any of our projects.

Reply:
current | remove entire section | make section project-specific |
--------|-----------------------|-------------------------------|
0 | 2 | 1 |

Reply:
current | remove entire section | make section project-specific
-- | -- | --
0 | 1 | 2

I changed my vote to the suggestion from @j-stephan

Reply:
I am fine with every project having their customizations. But I would remove the entire section for the clang-format file stored in this github repo.

Reply:
How about commenting it out? User an easily uncomment and adapt as needed.

Reply:
If it helps you, then just comment it out. I favor removal because it makes the file a few lines simpler. But honestly, I will probably never look at the clang-format file anyways, so the comments will not irritate me.

Reply:
What is the outcome of the vote now?

Reply:
Voting closed, the winner is "make it project-specific".